src/mlpack/methods/lmnn/constraints_impl.hpp

```diff
@@ -0,0 +1,148 @@
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_LMNN_CONSTRAINTS_IMPL_HPP
+#define MLPACK_METHODS_LMNN_CONSTRAINTS_IMPL_HPP
+
+// In case it hasn't been included already.
+#include "constraints.hpp"
+
+#include <mlpack/methods/neighbor_search/neighbor_search.hpp>
+
+namespace mlpack {
+namespace lmnn {
+
+Constraints::Constraints(
+    const arma::mat& dataset,
+    const arma::Row<size_t>& labels,
+    size_t k) :
+    dataset(math::MakeAlias(const_cast<arma::mat&>(dataset), false)),
+    labels(math::MakeAlias(const_cast<arma::Row<size_t>&>(labels), false)),
+    k(k)
+{
+  // Ensure a valid k is passed.
+  size_t minCount = arma::min(arma::histc(labels, arma::unique(labels)));
+
+  if (minCount < k)
+  {
+    Log::Fatal << "Constraints::VerifyK(): One of the class contains only "
+        << minCount << " instances, but value of k is " << k << "  "
+        << "(k should be < " << minCount << ")!" << std::endl;
+  }
+}
+
+// Calculates k similar labeled nearest neighbors.
+void Constraints::TargetNeighbors(arma::Mat<size_t>& outputMatrix)
+{
+  size_t N = dataset.n_cols;
+
+  arma::Row<size_t> uniqueLabels = arma::unique(labels);
+
+  outputMatrix = arma::Mat<size_t>(k, N, arma::fill::zeros);
+
+  // KNN instance.
+  neighbor::KNN knn;
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+
+  // Vectors to store indices.
+  arma::uvec indexSame;
+
+  for ( size_t i = 0; i < uniqueLabels.n_cols; i++)
+  {
+    // Calculate Target Neighbors.
+    indexSame = arma::find(labels == uniqueLabels[i]);
+
+    // Perform KNN search with same class points as both reference
+    // set and query set.
+    knn.Train(dataset.cols(indexSame));
+    knn.Search(k, neighbors, distances);
+
+    // Re-map neighbors to their index.
+    for (size_t j = 0; j < neighbors.n_elem; j++)
+      neighbors(j) = indexSame.at(neighbors(j));
+
+    // Store target neihbors.
+    outputMatrix.cols(indexSame) = neighbors;
+  }
+}
+
+// Calculates k differently labeled nearest neighbors.
+void Constraints::Impostors(arma::Mat<size_t>& outputMatrix)
+{
+  size_t N = dataset.n_cols;
+
+  arma::Row<size_t> uniqueLabels = arma::unique(labels);
+
+  outputMatrix = arma::Mat<size_t>(k, N, arma::fill::zeros);
+
+  // KNN instance.
+  neighbor::KNN knn;
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+
+  // Vectors to store indices.
+  arma::uvec indexSame;
+  arma::uvec indexDiff;
+
+  for ( size_t i = 0; i < uniqueLabels.n_cols; i++)
+  {
+    // Calculate impostors.
+    indexSame = arma::find(labels == uniqueLabels[i]);
+    indexDiff = arma::find(labels != uniqueLabels[i]);
+
+    // Perform KNN search with differently labeled points as reference
+    // set and  same class points as query set.
+    knn.Train(dataset.cols(indexDiff));
+    knn.Search(dataset.cols(indexSame), k, neighbors, distances);
```

Comment:
When we call `knn.Train()`, this builds a tree on the reference points.  That tree building process can take a long time, depending on the data.  (For instance, with MNIST it can take a while.)  However, KNN allows the use of a custom distance metric, and mlpack has the `MahalanobisDistance` metric in `src/mlpack/core/metrics/`.  So, it is almost possible to do something a little like this:

```
MahalanobisDistance md(dimensionality);
NeighborSearch<NearestNeighborSort, MahalanobisDistance> knn(referenceSet);
for (size_t i = 0; i < iterations; ++i)
{
  arma::mat parameters = GetParametersSomehow(); // just an example...
  knn.Metric().Covariance() = parameters;

  knn.Search(querySet, k, neighbors, distances);

  ...
}
```

Of course that is not the exact structure of the code that you have, but hopefully you see where I am going with it, which is that you can build the kd-tree once, then change the MahalanobisDistance covariance matrix, and get new results.

But there are some gotchas:

 * The `BinarySpaceTree<>` class also has a bounding structure.  By default `HRectBound` is used, but that will not work with `MahalanobisDistance`.  We would need to write a `MahalanobisHRectBound` class or something like this that used the distance matrix from the Mahalanobis distance.

 * As the Mahalanobis distance matrix gets further and further away from the identity matrix, the nearest neighbor search will become slower and slower as the bounds of the hyperrectangles in the tree change.  So at some point it becomes worthwhile to rebuild the tree.

 * You can also build the query tree only once to save some time.  By default each call to `Search()` will build a tree on the query points too, but you could pass in a pre-built tree.

This is kind of an in-depth optimization and would take a little time and implementation, but it could help speed things up.

Reply:
This could result in good optimization. Will have to think more about the code structure. I will update you as soon as I get something.

Reply:
Let's say we want to make this optimization to impostors part only (as we generally keep target neighbors constants except when shuffle is used, here also calling TargetNeighbors() can be avoided). So, if we want to employ this change we will need to store at least `uniqueLabels.n_elem` number of trees. Do you think this is efficient and do we have a way to do this?

Reply:
Each tree should have size somewhat less than 2x the dataset it is built on.  I think it's not likely that a user will manage to use a dataset that is so large that memory becomes an issue, since the optimization is sufficiently slow right now that I don't think we could do 100k points in any reasonable amount of time.  If we do accelerate it to that fast then maybe memory will become an issue then, but for now I wouldn't worry about it.